### ISSUE TITLE
Replace fails at larger datasets

### ISSUE THREAD

**Issue by lmocsi** [pola-rs/polars#12766]
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```
import polars as pl
n = 2000
df2 = pl.LazyFrame({"MIFID_FL": [0]*n + [1]*n})
df2.with_columns(pl.col('MIFID_FL').replace({0: 10, 1: 11})).select('MIFID_FL').group_by('MIFID_FL').count().collect()
```

### Log output

```shell
thread '<unnamed>' panicked at /home/runner/work/polars/polars/crates/polars-core/src/datatypes/dtype.rs:273:24:
internal error: entered unreachable code

---------------------------------------------------------------------------
PanicException                            Traceback (most recent call last)
Cell In[99], line 3
      1 n = 2000
      2 df2 = pl.LazyFrame({"MIFID_FL": [0]*n + [1]*n})
----> 3 df2.with_columns(pl.col('MIFID_FL').replace({0: 10, 1: 11})).select('MIFID_FL').group_by('MIFID_FL').count().collect()

File ~/.conda/envs/kh_spark32_env/lib/python3.9/site-packages/polars/utils/deprecation.py:100, in deprecate_renamed_parameter.<locals>.decorate.<locals>.wrapper(*args, **kwargs)
     95 @wraps(function)
     96 def wrapper(*args: P.args, **kwargs: P.kwargs) -> T:
     97     _rename_keyword_argument(
     98         old_name, new_name, kwargs, function.__name__, version
     99     )
--> 100     return function(*args, **kwargs)

File ~/.conda/envs/kh_spark32_env/lib/python3.9/site-packages/polars/lazyframe/frame.py:1788, in LazyFrame.collect(self, type_coercion, predicate_pushdown, projection_pushdown, simplify_expression, slice_pushdown, comm_subplan_elim, comm_subexpr_elim, no_optimization, streaming, _eager)
   1775     comm_subplan_elim = False
   1777 ldf = self._ldf.optimization_toggle(
   1778     type_coercion,
   1779     predicate_pushdown,
   (...)
   1786     _eager,
   1787 )
-> 1788 return wrap_df(ldf.collect())

PanicException: internal error: entered unreachable code
```

### Issue description

With n = 200 the code runs fine, returns the expected results.
With n = 2000 (and above), the above error is returned instead.

### Expected behavior

Should run fine with larger datasets, as well.

### Installed versions

<details>
--------Version info---------
Polars:               0.19.17
Index type:           UInt32
Platform:             Linux-5.10.198-187.748.amzn2.x86_64-x86_64-with-glibc2.26
Python:               3.9.18 | packaged by conda-forge | (main, Aug 30 2023, 03:49:32) 
[GCC 12.3.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          2.2.1
connectorx:           <not installed>
deltalake:            <not installed>
fsspec:               2023.10.0
gevent:               <not installed>
matplotlib:           <not installed>
numpy:                1.26.0
openpyxl:             <not installed>
pandas:               2.1.3
pyarrow:              14.0.1
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
</details>

**Comment by ritchie46** [comment]
The culprit is that `replace` gives a schema containing `Unkown`.

**Comment by lmocsi** [comment]
If I re-cast the output of replace() it seems to solve the problem, like:
   df2.with_columns(pl.col('MIFID_FL').replace({0: 10, 1: 11}).cast(pl.Int64)).select('MIFID_FL').group_by('MIFID_FL').count().collect()
runs fine.

**Comment by lmocsi** [comment]
> The culprit is that `replace` gives a schema containing `Unkown`.

Does this have to do anything with the phenomenon, that changing data type erases the field's contents?
Like this runs fine:
```
n = 200000
df2 = pl.LazyFrame({"MY_FL": ['N']*n + ['Y']*n})
df2.select('MY_FL').group_by('MY_FL').count().collect()
```

But this returns only nulls:
`df2.with_columns(pl.col('MY_FL').replace({'N': 0, 'Y': 1}).cast(pl.Int32)).select('MY_FL').group_by('MY_FL').count().collect()`

:-O

**Comment by ritchie46** [comment]
#12769 will ensure it will succeed on the larger dataset as well by not going into the streaming engine if the dtype is unknown. Though it still remains a logical bug (which would be great if we can fix it) if we don't know the output type.

**Comment by lmocsi** [comment]
Polars version 0.19.18 fixed the numeric example, but the string to number conversion issue is still there, see my last comment.

**Comment by cmdlineluser** [comment]
The string version now looks to be fixed on `0.20.0`

```python
df2.with_columns(
   pl.col('MY_FL').replace({'N': 0, 'Y': 1}).cast(pl.Int32)
).select('MY_FL').group_by('MY_FL').count().collect()

# shape: (2, 2)
# ┌───────┬────────┐
# │ MY_FL ┆ count  │
# │ ---   ┆ ---    │
# │ i32   ┆ u32    │
# ╞═══════╪════════╡
# │ 0     ┆ 200000 │
# │ 1     ┆ 200000 │
# └───────┴────────┘
```

**Comment by lmocsi** [comment]
Yes, I can confirm that with version 0.20 all the above issues are gone, working as expected. :)
Closing the  issue.
